### PR TITLE
fix(serializer): normalize style attributes

### PR DESCRIPTION
**Pull request by jmsjtu**
Split into two commits to make the changes more obvious.

See #236 for class changes.

BREAKING CHANGE: snapshots may change.

W-15365974

### CODE REVIEW

File: test/src/modules/serializer/styleAttr/__tests__/__snapshots__/styleAttr.spec.js.snap [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+      style="color:blue;text-align:center"
+    />
+    <div
+      style="color:blue; text-align:center"
+    />
+    <div
+      style="color: blue; text-align: center"
+    />
+    <div />
+    <div />
```
Comment (nolanlawson) [inline test/src/modules/serializer/styleAttr/__tests__/__snapshots__/styleAttr.spec.js.snap:23]: Let's add tests for `!important` and extra ending `;`s as well, and whitespace around each.

File: packages/@lwc/jest-serializer/src/clean-element-attrs.js [hunk 2]
```diff
@@ -23,6 +23,18 @@ function cleanElementAttributes(elm) {
...
+            // Normalize trivial whitespace differences (e.g. `style="  foo   bar "` becomes `style="foo bar"`
+            // and `style=""` is just removed).
+            const styleAttribute = elm.getAttribute(name) || '';
+            const normalizedStyleAttribute = styleAttribute.replace(/\s+/g, ' ').trim();
+            if (normalizedStyleAttribute) {
+                elm.setAttribute(name, normalizedStyleAttribute);
+            } else {
+                elm.removeAttribute(name);
+            }
+        }
```
Comment (nolanlawson) [inline packages/@lwc/jest-serializer/src/clean-element-attrs.js:37]: We should probably split on `;` and normalize whitespace for each, since there can be arbitrary whitespace before/after the semicolon.

Comment (nolanlawson) [reply packages/@lwc/jest-serializer/src/clean-element-attrs.js:37]: Also this would be cleaner to go in its own `cleanElementStyle` function/file imo.

Comment (nolanlawson) [reply packages/@lwc/jest-serializer/src/clean-element-attrs.js:37]: Also: this is not going to handle zany things like `content` or `grid-template-areas` which may contain spaces or semicolons in the value, but I'm not sure that that's worth handling. It quickly devolves into prettier/full CSS parsing, and we probably just need a lightweight formatter.

File: test/src/modules/serializer/styleAttr/__tests__/__snapshots__/styleAttr.spec.js.snap (hunk 1 above)
Comment (jmsjtu) [reply test/src/modules/serializer/styleAttr/__tests__/__snapshots__/styleAttr.spec.js.snap:66]: Added in the latest commit! 6d6e6ef557f0da3f8ecb578bc3921642caa4ec33

File: packages/@lwc/jest-serializer/src/clean-element-styles.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    const styleAttribute = elm.getAttribute('style') || '';
+    const normalizedStyleAttribute = styleAttribute
+        // Remove leading/trailing space
+        .trim()
+        // Replace multiple space, tab, line break with single space
+        .replace(/[\s\n\t]+/g, ' ')
+        .split(';')
+        // Preserve space character following a ';', style="color: blue; text-align: center"
+        .map((attr) => attr.trimEnd())
+        .join(';');
```
Comment (nolanlawson) [inline packages/@lwc/jest-serializer/src/clean-element-styles.js:20]:
```suggestion
        .split(';')
        .map(_ => _.trim())
        .filter(Boolean)
        .map(_ => `${_};`)
        .join(' ');
```

This avoids the duplicates `;`s at the end.

Comment (wjhsf) [reply packages/@lwc/jest-serializer/src/clean-element-styles.js:20]:
```suggestion
        .split(';')
        .map(_ => `${_.trim()};`)
        .filter(_ => _ !== ';')
        .join(' ');
```

Same but one less loop.

Comment (jmsjtu) [reply packages/@lwc/jest-serializer/src/clean-element-styles.js:20]: This will also append a ';' at the end of a style:

```html
<div style="  color: blue  "></div>
```

Becomes:

```html
<div style="color: blue;"></div>
```

Empty ';' will be removed as well

```html
<div style=";;"></div>
```
 Becomes:

```html
<div />
```

Just want to double-check, are we trying to normalize the whitespace or the style attribute output?

Asking because it won't exactly line up with what the component author will see in the browser.

Comment (nolanlawson) [reply packages/@lwc/jest-serializer/src/clean-element-styles.js]: Yes, I think we want consistent output. So consistent `;`s at the end, consistent spacing, and just `<div />` if the `style` is semantically empty.

File: test/src/modules/serializer/styleAttr/__tests__/__snapshots__/styleAttr.spec.js.snap [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+      style="color: blue; text-align: center;"
+    />
+    <div
+      style="color: blue; text-align: center;"
+    />
+    <div
+      style="color: blue; text-align: center;"
+    />
+    <div
+      style="color : blue; text-align : center;"
```
Comment (nolanlawson) [inline test/src/modules/serializer/styleAttr/__tests__/__snapshots__/styleAttr.spec.js.snap:33]: This one is going to lead to inconsistencies between static/dynamic sadly... Not sure how to solve without a full CSS parse 🤦

Comment (nolanlawson) [reply test/src/modules/serializer/styleAttr/__tests__/__snapshots__/styleAttr.spec.js.snap:33]: On second thought, we can't use a CSS parser because we've never validated this content, so it could be syntactically invalid and consumers could be relying on that.

I think we should split on `:` as well and normalize that so that it's always in the format:

```
color: red;
```

or

```
color: red !important;
```

Comment (nolanlawson) [reply test/src/modules/serializer/styleAttr/__tests__/__snapshots__/styleAttr.spec.js.snap:33]: #238

**Review by nolanlawson — APPROVED** [review]

**Review by nolanlawson — CHANGES_REQUESTED** [review]
See above and #239

**Review by nolanlawson — APPROVED** [review]
